gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/DataFlowTopology.java

```diff
@@ -0,0 +1,150 @@
+ * @author mitu
+ *
+ */
+@AllArgsConstructor
+public class DataFlowTopology {
+
+  public static final String ROUTES = "routes";
+  
+  @Getter
+  private final List<CopyRoute> routes;
+
+  private DataFlowTopology(Builder builder) {
+    Preconditions.checkArgument(builder.source != null, "Can not build topology without source");
+    Preconditions.checkArgument(builder.replicas != null && builder.replicas.size() > 0, "Can not build topology without replicas");
+    Preconditions.checkArgument(builder.topologyConfig != null, "Can not build topology without topology config");
+    Preconditions.checkArgument(builder.topologyConfig.hasPath(ROUTES), "Can not build topology without " + ROUTES);
+
+    // key of the map is replication name, value is {@link ReplicationReplica}
+    Map<String, ReplicationReplica> replicasMap = new HashMap<String, ReplicationReplica>();
+    for (ReplicationReplica replica : builder.replicas) {
+      String name = replica.getReplicationName();
+      // replica name can not be {@link ReplicationUtils#REPLICATION_SOURCE}
+      Preconditions.checkArgument(!name.equals(ReplicationUtils.REPLICATION_SOURCE),
+          String.format("replica name %s can not be reserved word %s ", name, ReplicationUtils.REPLICATION_SOURCE));
+      replicasMap.put(name, replica);
+    }
+
+    this.routes = new ArrayList<CopyRoute>();
+
+    Config routesConfig = builder.topologyConfig.getConfig(ROUTES);
+    // routes should be available for each replica
+    for (ReplicationReplica replica : builder.replicas) {
+      Preconditions.checkArgument(routesConfig.hasPath(replica.getReplicationName()),
+          "can not find route for replica " + replica.getReplicationName());
+
+      List<ReplicationData<CopyableDataset>> copyFromReplica = new ArrayList<ReplicationData<CopyableDataset>>();
+
+      List<String> copyFromStrings = routesConfig.getStringList(replica.getReplicationName());
+
+      for (String copyFromName : copyFromStrings) {
+        Preconditions.checkArgument(!copyFromName.equals(replica.getReplicationName()),
+            "can not have same name in both destination and copy from list " + copyFromName);
+        // copy from source
+        if (copyFromName.equals(ReplicationUtils.REPLICATION_SOURCE)) {
+          copyFromReplica.add(builder.source);
+        }
+        // copy from other replicas
+        else if(replicasMap.containsKey(copyFromName)){
+          copyFromReplica.add(replicasMap.get(copyFromName));
+        }
+        else{
+          throw new IllegalArgumentException("can not find replica with name " + copyFromName);
+        }
+      }
+      
+      CopyRouteBuilder routeBuilder = new CopyRouteBuilder();
+      routeBuilder.withCopyDestination(replica);
+      routeBuilder.withCopyFrom(copyFromReplica);
+      this.routes.add(routeBuilder.build());
+    }
+  }
+
+  public static class CopyRoute {
+    @Getter
+    private final ReplicationReplica copyDestination;
```

Comment:
nit: it's nice to have the names be consistent: either copyFrom/copyTo or copySource/copyDestination or routeSource/routeDestination.